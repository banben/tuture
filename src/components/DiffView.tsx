import React, { Component } from 'react';
import { injectGlobal } from 'styled-components';
import classnames from 'classnames';

import Snippet from './Snippet';

import { Change as ChangeType, Hunk as HunkType } from '../types';

interface DiffViewProps {
  startLine: number;
  hunks: HunkType[];
}

injectGlobal`
  .diff {
    table-layout: fixed;
    border-collapse: collapse;
    width: 100%;
  }

  .diff-gutter-col {
    width: 50px;
  }

  .diff td {
    vertical-align: top;
  }
  .diff-line {
    line-height: 31px;
    font-family: Consolas, Courier, monospace;
  }

  .diff-gutter {
    width: 8px;
    padding: 0 16px;
    color: rgba(0, 0, 0, .24);
  }

  .diff-gutter-insert {
    background-color: rgba(0, 0, 0, .07);
  }

  .diff-gutter-delete {
    background-color: rgba(0, 0, 0, .021);
  }

  .diff-gutter:empty:before {
    content: attr(data-line-number);
  }

  .diff-code-insert {
    font-weight: 700;
    background-color: rgba(0, 0, 0, .07);
  }

  .diff-code-delete {
    opacity: 0.3;
    background-color: rgba(0, 0, 0, .07);
  }

  .diff-code {
    padding: 0 20px;
    width: 557px;
  }
  .diff-file {
    color: rgba(0,0,0,0.84);
    display: block;
    padding-top: 8px;
    padding-bottom: 20px;
    background-color: rgba(0, 0, 0, .05);
    margin: 32px 0;
  }

  .diff-file-header {
    font-family: Monaco;
    font-size: 14px;
    color: rgba(0,0,0,0.24);
    text-align: right;
    padding-right: 20px;
  }

  .addition-count {
    margin-right: 1em;
    color: #88b149;
  }

  .deletion-count {
    margin-right: 1em;
    color: #ee5b60;
  }

`;

export default class DiffView extends Component<DiffViewProps> {
  renderLineNumber = (
    lineNumberClassName: string,
    lineNumber: number,
  ): React.ReactNode => {
    return <td className={lineNumberClassName} data-line-number={lineNumber} />;
  };

  renderRow = (change: ChangeType, isAllInsert: Boolean, i: number) => {
    const { type, content } = change;
    const lineNumberClassName = classnames('diff-gutter', {
      [`diff-gutter-${type}`]: !isAllInsert,
    });
    const codeClassName = classnames('diff-code', {
      [`diff-code-${type}`]: !isAllInsert,
    });
    return (
      <tr key={`change${i}`} className={classnames('diff-line')}>
        {this.renderLineNumber(lineNumberClassName, this.props.startLine + i)}
        <td className={codeClassName}>
          <Snippet code={content} />
        </td>
      </tr>
    );
  };

  judgeAllRowInsertState = (changes: ChangeType[]) => {
    let isAllInsert = true;
    changes.map((change: ChangeType) => {
      const { type } = change;
      if (type !== 'insert') {
        isAllInsert = false;
      }
    });
    return changes.map((change, i) => {
      return this.renderRow(change, isAllInsert, i);
    });
  };

  render() {
    const { hunks } = this.props;

    return (
      <table className="diff">
        {hunks.map((hunk: HunkType, key: number) => (
          <tbody key={key} className={classnames('diff-hunk')}>
            {this.judgeAllRowInsertState(hunk.changes)}
          </tbody>
        ))}
      </table>
    );
  }
}
