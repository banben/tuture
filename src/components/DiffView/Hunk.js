import React, { Component } from 'react';

import UnifiedHunk from './UnifiedHunk';
import SplitHunk from './SplitHunk';

import './css/Change.css';

export default class Hunk extends Component {
  render() {
    const { viewType, hunk } = this.props;
    const RenderingHunk = viewType === 'unified' ? UnifiedHunk : SplitHunk;

    return (
      <RenderingHunk
        hunk={hunk}
      />
    );
  }
}
