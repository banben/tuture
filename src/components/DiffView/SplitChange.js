import React, { PureComponent } from 'react';
import classnames from 'classnames';

import {
  computeOldLineNumber,
  computeNewLineNumber,
} from './utils';

export default class SplitChange extends PureComponent {
  SIDE_OLD = 0;
  SIDE_NEW = 1;

  renderCells = ({
    side,
    change,
    gutterAnchor,
    hideGutter,
    gutterAnchorTarget
  }) => {
    if (!change) {
      const gutterClassName = classnames('diff-gutter', 'diff-gutter-omit');
      const codeClassName = classnames('diff-code', 'diff-code-omit');

      return [
        !hideGutter && <td key="gutter" className={gutterClassName} />,
        <td key="code" className={codeClassName} />,
      ];
    }

    const { type, content } = change;
    const line = side === this.SIDE_OLD
      ? computeOldLineNumber(change)
      : computeNewLineNumber(change)
    ;
    const gutterClassName = classnames(
      'diff-gutter',
      `diff-gutter-${type}`,
    );
    const gutterProps = {
      className: gutterClassName,
      'data-line-number': line,
      children: (
        gutterAnchor
        ? <a href={`#${gutterAnchorTarget}`} data-line-number={line} />
        : null
      ),
    };
    const codeClassName = classnames(
      'diff-code',
      `diff-code-${type}`,
    );
    const codeProps = {
      className: codeClassName,
    };

    return [
      !hideGutter && <td key="gutter" {...gutterProps} />,
      <td key="code" {...codeProps}>{content}</td>
    ];
  }

  render() {
    const {
      oldChange,
      newChange,
    } = this.props;

    const oldArgs = {
      change: oldChange,
      side: this.SIDE_OLD,
    };
    const newArgs = {
      change: newChange,
      side: this.SIDE_NEW,
    };

    const lineTypeClassName = ((oldChange, newChange) => {
      if (oldChange && !newChange) {
        return 'diff-line-old-only';
      }

      if (!oldChange && newChange) {
        return 'diff-line-new-only';
      }

      if (oldChange === newChange) {
        return 'diff-line-normal';
      }

      return 'diff-line-compare';
    })(oldChange, newChange);

    return (
      <tr className={classnames('diff-line')}>
        {this.renderCells(oldArgs)}
        {this.renderCells(newArgs)}
      </tr>
    );
  }
}
