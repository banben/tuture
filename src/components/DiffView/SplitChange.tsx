import React, { PureComponent } from 'react';
import classnames from 'classnames';

import {
  computeOldLineNumber,
  computeNewLineNumber,
} from './utils';

import { Change as ChangeType } from '../ContentItem';

export interface SplitChangeProps {
  oldChange: ChangeType;
  newChange: ChangeType;
}

export interface Cell {
  side?: number;
  change?: ChangeType;
  gutterAnchor?: string;
  hideGutter?: boolean;
  gutterAnchorTarget?: string;
  anchorID?: number;
  lineNumber?: number;
  hide?: boolean;
  className?: string;
}

export default class SplitChange extends PureComponent<SplitChangeProps> {
  constructor(props: SplitChangeProps) {
    super(props);
  }

  SIDE_OLD = 0;
  SIDE_NEW = 1;

  renderCells = ({
    side,
    change,
    gutterAnchor,
    hideGutter,
    gutterAnchorTarget
  }: Cell): React.ReactNode => {
    if (!change) {
      const gutterClassName = classnames('diff-gutter', 'diff-gutter-omit');
      const codeClassName = classnames('diff-code', 'diff-code-omit');

      return [
        !hideGutter && <td key="gutter" className={gutterClassName} />,
        <td key="code" className={codeClassName} />,
      ];
    }

    const { type, content } = change;
    const line = side === this.SIDE_OLD
      ? computeOldLineNumber(change)
      : computeNewLineNumber(change)
    ;
    const gutterClassName = classnames(
      'diff-gutter',
      `diff-gutter-${type}`,
    );
    const gutterProps = {
      className: gutterClassName,
      'data-line-number': line,
      children: (
        gutterAnchor
        ? <a href={`#${gutterAnchorTarget}`} data-line-number={line} />
        : null
      ),
    };
    const codeClassName = classnames(
      'diff-code',
      `diff-code-${type}`,
    );
    const codeProps = {
      className: codeClassName,
    };

    return [
      !hideGutter && <td key="gutter" {...gutterProps} />,
      <td key="code" {...codeProps}>{content}</td>
    ];
  }

  render() {
    const {
      oldChange,
      newChange,
    } = this.props;

    const oldArgs = {
      change: oldChange,
      side: this.SIDE_OLD,
    };
    const newArgs = {
      change: newChange,
      side: this.SIDE_NEW,
    };

    const lineTypeClassName = ((oldChange, newChange) => {
      if (oldChange && !newChange) {
        return 'diff-line-old-only';
      }

      if (!oldChange && newChange) {
        return 'diff-line-new-only';
      }

      if (oldChange === newChange) {
        return 'diff-line-normal';
      }

      return 'diff-line-compare';
    })(oldChange, newChange);

    return (
      <tr className={classnames('diff-line')}>
        {this.renderCells(oldArgs)}
        {this.renderCells(newArgs)}
      </tr>
    );
  }
}
