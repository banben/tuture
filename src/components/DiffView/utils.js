function computeOldLineNumber(change) {
  if (change.isInsert) {
    return -1;
  }

  return change.isNormal ? change.oldLineNumber : change.lineNumber;
}

function computeNewLineNumber(change) {
  if (change.isDelete) {
    return -1;
  }

  return change.isNormal ? change.newLineNumber : change.lineNumber;
}

function getChangeKey(change) {
  if (!change) {
    throw new Error('change is not provided');
  }

  const {
    isNormal,
    isInsert,
    lineNumber,
    oldLineNumber,
  } = change;

  if (isNormal) {
    return 'N' + oldLineNumber;
  }

  const prefix = isInsert ? 'I' : 'D';
  return prefix + lineNumber;
}

export {
  computeOldLineNumber,
  computeNewLineNumber,
  getChangeKey,
}
