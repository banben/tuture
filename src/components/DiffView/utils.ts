import { Change as ChangeType } from '../../types';

function computeOldLineNumber(change: ChangeType): number {
  if (change.isInsert) {
    return -1;
  }

  return change.isNormal ? change.oldLineNumber : change.lineNumber;
}

function computeNewLineNumber(change: ChangeType): number {
  if (change.isDelete) {
    return -1;
  }

  return change.isNormal ? change.newLineNumber : change.lineNumber;
}

function getChangeKey(change: ChangeType): string {
  if (!change) {
    throw new Error('change is not provided');
  }

  const {
    isNormal,
    isInsert,
    lineNumber,
    oldLineNumber,
  } = change;

  if (isNormal) {
    return 'N' + oldLineNumber;
  }

  const prefix = isInsert ? 'I' : 'D';
  return prefix + lineNumber;
}

export {
  computeOldLineNumber,
  computeNewLineNumber,
  getChangeKey,
}
