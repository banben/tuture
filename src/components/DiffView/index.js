import React, { Component } from 'react';

import Hunk from './Hunk';

export default class DiffView extends Component {
  render() {
    const { viewType, hunks } = this.props;

    const gutterContent = (
      viewType === 'unified'
      ? (
        <colgroup>
          <col className="diff-gutter-col"/>
          <col className="diff-gutter-col"/>
        </colgroup>
      ) : (
        <colgroup>
          <col className="diff-gutter-col"/>
          <col/>
          <col className="diff-gutter-col"/>
          <col/>
        </colgroup>
      )
    )

    return (
      <table className="diff">
        {gutterContent}
        {
          hunks.map((hunk, key) => (
            <Hunk key={key} viewType={viewType} hunk={hunk} />
          ))
        }
      </table>
    );
  }
}
