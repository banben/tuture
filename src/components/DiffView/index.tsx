import React, { Component } from 'react';

import Hunk from './Hunk';

import { Hunk as HunkType } from '../ContentItem';

interface DiffViewProps {
  viewType: string;
  hunks: HunkType[];
}

export default class DiffView extends Component<DiffViewProps> {
  render() {
    const { viewType, hunks } = this.props;

    const gutterContent = (
      viewType === 'unified'
      ? (
        <colgroup>
          <col className="diff-gutter-col"/>
          <col className="diff-gutter-col"/>
        </colgroup>
      ) : (
        <colgroup>
          <col className="diff-gutter-col"/>
          <col/>
          <col className="diff-gutter-col"/>
          <col/>
        </colgroup>
      )
    )

    return (
      <table className="diff">
        {gutterContent}
        {
          hunks.map((hunk: HunkType, key: number) => (
            <Hunk key={key} viewType={viewType} hunk={hunk} />
          ))
        }
      </table>
    );
  }
}
