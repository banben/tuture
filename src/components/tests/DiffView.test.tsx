import React from 'react';
import renderer from 'react-test-renderer';

import DiffView from '../DiffView/index';
import { hunk } from './utils/hunData';
import { Hunk } from '../../types/index';

test('Test DiffView render successfully', () => {
  let hunks: Hunk[]  = [];
  Array(10).map(item => hunks.push(hunk));

  const component = renderer.create(
    <DiffView
      hunks={hunks}
      viewType={'unified'}
      key={'jus for fun'}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
