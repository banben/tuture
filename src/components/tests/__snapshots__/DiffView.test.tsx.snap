// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DiffView render successfully 1`] = `
<table
  className="diff"
>
  <colgroup>
    <col
      className="diff-gutter-col"
    />
    <col
      className="diff-gutter-col"
    />
  </colgroup>
</table>
`;
